cpp/include/cudf/io/parquet.hpp

```diff
@@ -0,0 +1,617 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @file parquet.hpp
+ * @brief cuDF-IO freeform API
+ */
+
+#pragma once
+
+#include <cudf/io/types.hpp>
+#include <iostream>
+
+#include <cudf/io/writers.hpp>
+#include <cudf/table/table_view.hpp>
+#include <cudf/types.hpp>
+
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+//! cuDF interfaces
+namespace cudf {
+//! In-development features
+namespace io {
+
+/**
+ * @brief Builds parquet_reader_options to use for `read_parquet()`
+ */
+class parquet_reader_options_builder;
+
+/**
+ * @brief Settings to use for `read_parquet()`
+ */
+class parquet_reader_options {
+  source_info _source;
+
+  /// Names of column to read; empty is all
+  std::vector<std::string> _columns;
+
+  /// List of individual row groups to read (ignored if empty)
+  std::vector<std::vector<size_type>> _row_groups;
+  /// Rows to skip from the start; -1 is none
+  size_type _skip_rows = -1;
+  /// Rows to read; -1 is all
+  size_type _num_rows = -1;
+
+  /// Whether to store string data as categorical type
+  bool _strings_to_categorical = false;
+  /// Whether to use PANDAS metadata to load columns
+  bool _use_pandas_metadata = true;
+  /// Cast timestamp columns to a specific type
+  data_type _timestamp_type{type_id::EMPTY};
+
+  explicit parquet_reader_options(source_info const& src) : _source(src) {}
+
+  friend parquet_reader_options_builder;
+
+ public:
+  explicit parquet_reader_options() = default;
+
+  /**
+   * @brief enum class for parquet_reader_options boolean parameters
+   */
+
+  enum class boolean_param_id : int8_t {
+    STRINGS_TO_CATEGORICAL,  // Whether to store string data as categorical type
+    USE_PANDAS_METADATA,     // Whether to use PANDAS metadata to load columns
+  };
+
+  /**
+   * @brief enum class for parquet_reader_options size_type parameters
+   */
+  enum class size_type_param_id : int8_t {
+    SKIP_ROWS,  // Rows to skip from the start
+    NUM_ROWS    // Rows to read
+  };
+
+  /**
+   * @brief create parquet_reader_options_builder which will build parquet_reader_options
+   *
+   * @param src source information used to read parquet file
+   * @returns parquet_reader_options_builder
+   */
+  static parquet_reader_options_builder builder(source_info const& src);
+
+  /**
+   * @brief Returns source info
+   */
+  source_info const& source() const { return _source; }
+
+  /**
+   * @brief Returns boolean parameter values as per the corresponding enum
+   */
+  bool get(boolean_param_id param_id) const
+  {
+    switch (param_id) {
+      case boolean_param_id::STRINGS_TO_CATEGORICAL: return _strings_to_categorical;
+      case boolean_param_id::USE_PANDAS_METADATA: return _use_pandas_metadata;
+      default: CUDF_FAIL("Unsupported boolean_param_id enum");
+    }
```

Comment:
Couldn't you store a std::map of ID->bool so that adding a new parameter is only a matter of adding it to the enum and defining a default value?

Or an array of bool (with some casting shenanigans due to enum class).

Reply:
I was little hesitant regarding creating a map/array, it would be nice to observe all data members underneath the class rather than have it wrapped under a map